File: app/src/main/java/com/sergiigalai/chargetimer/charge/LiionChargeTimeResolver.java
package com.sergiigalai.chargetimer.charge;

public class LiionChargeTimeResolver implements IChargeTimeResolver
{
    private static final short LINEAR_CHARGING_THRESHOLD_PCT = 80;
    private PowerLine powerLine;
    private Battery battery;

    public LiionChargeTimeResolver(PowerLine powerLine, Battery battery) {
        this.powerLine = powerLine;
        this.battery = battery;
    }

    @Override
    public long getMillisToCharge(byte remainingEnergyPct){
        double hoursToCharge;

        if (remainingEnergyPct < LINEAR_CHARGING_THRESHOLD_PCT){
            hoursToCharge = getLinearDependencyTime(LINEAR_CHARGING_THRESHOLD_PCT, remainingEnergyPct);
            hoursToCharge += getFinishChargingTime(LINEAR_CHARGING_THRESHOLD_PCT);
        } else {
            hoursToCharge = getFinishChargingTime(remainingEnergyPct);
        }

        return convertHoursToMs(hoursToCharge);
    }

    private long convertHoursToMs(double hours) {
        Double ms = hours * 3600 * 1000;
        return ms.isInfinite() ? 0 : ms.longValue();
    }

    private int getChargingPowerWh() {
        return powerLine.Voltage * powerLine.Amperage;
    }

    private double getChargingEfficiency() {
        return 100 / (100 + battery.ChargingLossPct);
    }

    private double getLinearDependencyTime(short maxPercentage, short currentPercentage) {
        return getEmptyBatteryChargeTime() *
                (maxPercentage - currentPercentage) / 100;
    }

    private double getFinishChargingTime(int currentPercentage) {
        final int ChargingAmperageSlowdown = 2;

        return getEmptyBatteryChargeTime() * ChargingAmperageSlowdown *
                (100 - currentPercentage) / 100;
    }

    private double getEmptyBatteryChargeTime() {
        double WhToCharge = battery.UsableCapacityKWh * 1000;
        return WhToCharge / (getChargingPowerWh() * getChargingEfficiency());
    }
}

File: EvChargeTimer/app/src/main/java/com/chebuso/chargetimer/charge/LiIonChargeTimeResolver.kt
package com.chebuso.chargetimer.charge

class LiIonChargeTimeResolver(private val powerLine: PowerLine,
                              private val battery: Battery)
{
    fun getTimeToCharge(): Long {
        var hoursToCharge: Double
        if (battery.remainingEnergyPct < LINEAR_CHARGING_THRESHOLD_PCT) {
            hoursToCharge =
                getLinearDependencyTime(LINEAR_CHARGING_THRESHOLD_PCT, battery.remainingEnergyPct.toShort())
            hoursToCharge += getFinishChargingTime(LINEAR_CHARGING_THRESHOLD_PCT.toInt())
        } else {
            hoursToCharge = getFinishChargingTime(battery.remainingEnergyPct.toInt())
        }
        return convertHoursToMs(hoursToCharge)
    }

    private fun convertHoursToMs(hours: Double): Long {
        val ms = hours * 3600 * 1000
        return if (ms.isInfinite()) 0 else ms.toLong()
    }

    private val chargingPowerWh: Int
        get() = powerLine.Voltage * powerLine.Amperage

    private val chargingEfficiency: Double
        get() = 100 / (100 + battery.ChargingLossPct)

    private fun getLinearDependencyTime(maxPercentage: Short, currentPercentage: Short): Double {
        return emptyBatteryChargeTime *
                (maxPercentage - currentPercentage) / 100
    }

    private fun getFinishChargingTime(currentPercentage: Int): Double {
        val chargingAmperageSlowdown = 2
        return emptyBatteryChargeTime * chargingAmperageSlowdown *
                (100 - currentPercentage) / 100
    }

    private val emptyBatteryChargeTime: Double
        get() {
            val whToCharge: Double = battery.UsableCapacityKWh * 1000
            return whToCharge / (chargingPowerWh * chargingEfficiency)
        }

    companion object {
        private const val LINEAR_CHARGING_THRESHOLD_PCT: Short = 80
    }
}


